api/v1alpha1/common_types.go

```diff
@@ -14,10 +14,35 @@
 package v1alpha1
 
 import (
+	"math"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+// ProtectorSpec protects the target cluster from being corrupted by the chaos.
+// For now, some chaos could not be recovered such as pod-kill. If we continue
+// to inject it into our application cluster, it will finally pull down our
+// entire application cluster. That may be not our intention. We should protect
+// our application cluster from that disaster via a protector to make the
+// application cluster always in available.
+type ProtectorSpec struct {
+
+	// MaxInjectPercent specifies the ratio of maximum pods involved in the
+	// chaos to be injected later.  A value of less than or equal to 0, or
+	// greater than 100, we see it as illegal. It takes no effects for the
+	// protection.
+	MaxInjectPercent int `json:"maxInjectPercent,omitempty"`
```

Comment:
I suggest replace the `MaxInjectPercent` to the `MinUnInfectedCount`.  The uninfected replica counts would be flexible during each round of syncing considering the `pod-kill` case or the external factor while the `MinUnInfectedCount` would always guarantee the expected health instances.

Reply:
I think use `MinUnInfectedCount` is another point of view. Maybe we can support both at the same time ? @Yisaer

Reply:
About ** considering the pod-kill case**, I did not fully understand this.I guess that in each round of a `pod-kill` action the selected pods will be decreased, however the `MaxInjectPercent` should calculated according to the number of initial selected pods. The implementation did that exactly right now. For example, we set `MaxInjectPercent` to `30`. If we selected 10 pods in the first reconciling round of the `pod-kill`, the max injected pods should always less than or equal to `10*0.3` in the later rounds. @Yisaer

Reply:
Things would be much complex if we use radio here. And I also don't recommend recording the max injected pod counts in the first reconciling, considering not only the scaling (especially elastic auto-scaling) is common to see in the Kubernetes, (If the total pods which controlled by deployment have 10 replicas at initial, then we recorded 3 as max injected pods, and after that, the replicas of the deployment become 2, and the chaos-mesh kill all the pods finally), but also this design would make chaos-mesh controller more and more stateful. (I didn't see the benefit to use radio here for solving #179 comparing the fixed uninfected pods strategy)

Reply:
Agreed ! I missed that services will be scaled during chaos injection. A stateless service is more preferred.